handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -0,0 +1,1951 @@
+    }
+
+    @Override
+    public final synchronized void closeOutbound() {
+        if (isOutboundDone) {
+            return;
+        }
+
+        isOutboundDone = true;
+        engineClosed = true;
+
+        if (handshakeState != HandshakeState.NOT_STARTED && !isDestroyed()) {
+            int mode = SSL.getShutdown(ssl);
+            if ((mode & SSL.SSL_SENT_SHUTDOWN) != SSL.SSL_SENT_SHUTDOWN) {
+                int err = SSL.shutdownSSL(ssl);
+                if (err < 0) {
+                    int sslErr = SSL.getError(ssl, err);
+                    switch (sslErr) {
+                        case SSL.SSL_ERROR_NONE:
+                        case SSL.SSL_ERROR_WANT_ACCEPT:
+                        case SSL.SSL_ERROR_WANT_CONNECT:
+                        case SSL.SSL_ERROR_WANT_WRITE:
+                        case SSL.SSL_ERROR_WANT_READ:
+                        case SSL.SSL_ERROR_WANT_X509_LOOKUP:
+                        case SSL.SSL_ERROR_ZERO_RETURN:
+                            // Nothing to do here
+                            break;
+                        case SSL.SSL_ERROR_SYSCALL:
+                        case SSL.SSL_ERROR_SSL:
+                            if (logger.isDebugEnabled()) {
+                                logger.debug("SSL_shutdown failed: OpenSSL error: {}", SSL.getLastError());
+                            }
+                            // There was an internal error -- shutdown
+                            shutdown();
+                            break;
+                        default:
+                            SSL.clearError();
+                            break;
+                    }
+                }
+            }
+        } else {
+            // engine closing before initial handshake
+            shutdown();
+        }
+    }
+
+    @Override
+    public final synchronized boolean isOutboundDone() {
+        return isOutboundDone;
+    }
+
+    @Override
+    public final String[] getSupportedCipherSuites() {
+        return OpenSsl.AVAILABLE_CIPHER_SUITES.toArray(new String[OpenSsl.AVAILABLE_CIPHER_SUITES.size()]);
+    }
+
+    @Override
+    public final String[] getEnabledCipherSuites() {
+        final String[] enabled;
+        synchronized (this) {
+            if (!isDestroyed()) {
+                enabled = SSL.getCiphers(ssl);
+            } else {
+                return EmptyArrays.EMPTY_STRINGS;
+            }
+        }
+        if (enabled == null) {
+            return EmptyArrays.EMPTY_STRINGS;
+        } else {
+            synchronized (this) {
+                for (int i = 0; i < enabled.length; i++) {
+                    String mapped = toJavaCipherSuite(enabled[i]);
+                    if (mapped != null) {
+                        enabled[i] = mapped;
+                    }
+                }
+            }
+            return enabled;
+        }
+    }
+
+    @Override
+    public final void setEnabledCipherSuites(String[] cipherSuites) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)